---
title: "Microbiome Analysis - Introduction"
author: "Sehyun Oh"
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
      highlight: pygments
      toc: true
      toc_float: false
      toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE, collapse=TRUE)
```

# Introduction

Bioconductor provides curated resources of microbiome data. Most microbiome data 
are generated either by targeted amplicon sequencing (usually of variable regions 
of the 16S ribosomal RNA gene) or by metagenomic shotgun sequencing (MGX). These 
two approaches are analyzed by different sequence analysis tools, but downstream 
statistical and ecological analysis can involve any of the following types of data:

* taxonomic abundance at different levels of the taxonomic hierarchy
* phylogenetic distances and the phylogenetic tree of life
* metabolic potential of the microbiome
* abundance of microbial genes and gene families

A review of types and properties of microbiome data is provided by [@Morgan2012-mq].

# curatedMetagenomicData 
## Curated and processed metagenomic data through ExperimentHub

```{r results="hide", echo=FALSE, cache=FALSE, eval=TRUE}
suppressPackageStartupMessages(library(curatedMetagenomicData))
```

`curatedMetagenomicData`[@Pasolli2017-gf] provides 6 types of processed data for >30 
publicly available whole-metagenome shotgun sequencing datasets, including from the 
Human Microbiome Project Phase 1:

1. Species-level taxonomic profiles, expressed as relative abundance from kingdom to strain level
2. Presence of unique, clade-specific markers
3. Abundance of unique, clade-specific markers
4. Abundance of gene families
5. Metabolic pathway coverage
6. Metabolic pathway abundance

Types 1-3 are generated by [MetaPhlAn2](http://huttenhower.sph.harvard.edu/metaphlan2); 
4-6 are generated by [HUMAnN2](http://huttenhower.sph.harvard.edu/humann2).

Currently, `curatedMetagenomicData` provides:

* `r nrow(combined_metadata)` samples from `r length(unique(combined_metadata$dataset_name))` datasets, primarily of the human gut but including body sites profiled in the Human Microbiome Project
* Processed data from whole-metagenome shotgun metagenomics, with manually-curated metadata, as integrated and documented Bioconductor ExpressionSet objects
* ~80 fields of specimen metadata from original papers, supplementary files, and websites, with manual curation to standardize annotations
* Processing of data through the [MetaPhlAn2](http://huttenhower.sph.harvard.edu/metaphlan2) pipeline for taxonomic abundance, and [HUMAnN2](http://huttenhower.sph.harvard.edu/humann2) pipeline for metabolic analysis
* These represent ~100TB of raw sequencing data, but the processed data provided are much smaller.

These datasets are documented in the 
[reference manual](https://bioconductor.org/packages/devel/data/experiment/manuals/curatedMetagenomicData/man/curatedMetagenomicData.pdf).

## `curatedMetagenomicData`
This is an `ExperimentHub` package, and its main workhorse function is `curatedMetagenomicData()`:
The manually curated metadata for all available samples are provided in a single table `combined_metadata`:

```{r eval=FALSE}
library(curatedMetagenomicData)
?combined_metadata
View(data.frame(combined_metadata))
```

The main function provides a `list` of `ExpressionSet` objects:
```{r results="hide"}
oral <- c("HMP_2012.metaphlan_bugs_list.oralcavity")
esl <- curatedMetagenomicData(oral, dryrun = FALSE)
```

```{r}
esl
```

## `ExpressionSet2phyloseq`
These `ExpressionSet` objects can also be converted to `phyloseq` object for ecological analysis 
and differential abundance analysis using the `DESeq2` package, using the `ExpressionSet2phyloseq()` function:
```{r}
ExpressionSet2phyloseq(esl[[1]], phylogenetictree = TRUE)
```

See the documentation of `phyloseq` for more on ecological and differential abundance analysis of the microbiome.



# HMP16SData
## 16S rRNA Sequencing Data from the Human Microbiome Project
```{r results='hide'}
suppressPackageStartupMessages(library(HMP16SData))
```

`HMP16SData`[@Schiffer2019-cd] is a Bioconductor ExperimentData package of the Human Microbiome Project (HMP) 16S rRNA sequencing data. Taxonomic count data files are provided as downloaded from the HMP Data Analysis and Coordination Center from its QIIME pipeline. Processed data is provided as `SummarizedExperiment` class objects via `ExperimentHub`. Like other ExperimentHub-based packages, a convenience function does downloading, automatic local caching, and serializing of a Bioconductor data class. This returns taxonomic counts from the V1-3 variable region of the 16S rRNA gene, along with the unrestricted participant data and phylogenetic tree. 

```{r}
V13()
```

This can also be converted to `phyloseq` for ecological and differential abundance analysis; see the `HMP16SData` vignette for details.




# HMP2Data
## Integrative Human Microbiome Project (iHMP) (https://hmpdacc.org/ihmp/)
- One of the largest open data resources for studying longitudinal microbiome changes and connection to disease
- Novel data -- first scientific results just published  in the special collection 
of **Nature** and its sister journals (https://www.nature.com/articles/s41586-019-1238-8) 
```{r out.width='100%', out.height='70%', results='hold', echo=FALSE, eval=FALSE}
knitr::include_graphics(system.file(package='MicrobiomeWorkshop', 'vignettes', "iHMP.jpeg"))
```

- Human Microbiome Project Data Portal (https://portal.hmpdacc.org/)
- Open data, but not straightforward to download and work with
```{r out.width='80%', out.height='50%', results='hold', echo=FALSE, eval=FALSE}
knitr::include_graphics(system.file(package='MicrobiomeWorkshop', 'vignettes', "Nature.jpeg"))
```

```{r out.width='100%', out.height='70%', results='hold', echo=FALSE, eval=FALSE}
knitr::include_graphics(system.file(package='MicrobiomeWorkshop', 'vignettes', "DAC.jpeg"))
```

- Need **Aspera client** to even download the data  -- not every lab/researcher has these expertise available   
- Still need multiple processing steps after data is downloaded to:   
1. map taxonomy ids from the database   
2. merge with meta data -- not available on the DAC portal   
3. construct phylogenetic tree   
4. merge omics data
