---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
suppressPackageStartupMessages({
  library(AnVIL)
})
```

## Inputs for this workflow
```{r}
fpath <- "gs://fc-07ee4ddc-5b5b-46f6-bed7-809aa14bb012/IBDMDB/ibdmdb_file_list.txt"
inputs <- read.table(gsutil_pipe(fpath), sep = "\t")
inputs
```

Extract sample names from input file names
```{r}
sampleName <- gsub("_R1.fastq.gz", "", basename(inputs$V1))
```


## Setup correct workspace
```{r}
avworkspace_namespace("waldronlab-terra-rstudio")
avworkspace_name("mtx_workflow_biobakery_ver3")
avbucket()
```

## Collect output `.tsv` files
It seems like that `avworkflow_files` function accesses the most recent job. I can
specify a specific submission, which can be found using `avworkflow_jobs()`.

```{r}
avworkflow_jobs()
outputs <- avworkflow_files()
ind <- grep(".tsv", outputs$file)
outputs$file[ind]
```

### Available outputs from the current workflow
- genefamilies_relab   
- pathabundance_relab    
- pathcoverage    

Copy workflow outputs in a file system to apply `make_data`
```{r}
source("/home/rstudio/MicrobiomeWorkshop/inst/scripts/Biobakery_to_ExpressionSet.R")

# [1] "No output is available for marker_abundance"
# [1] "No output is available for marker_presence"
# [1] "No output is available for metadata"
# [1] "No output is available for metaphlan_bugs_list"
```


## Manually apply `make_data` steps
```{r eval=FALSE}
# make_data <- function(tar_gz_file, tmpdir=cmg_tmpdir() ) {
#     dir.create(uploads_dir(path=tmpdir))
#     dsname = gsub(".tar.gz", "", basename(tar_gz_file))
#     untar(tar_gz_file, exdir = tmpdir)
#     dataset_dir <- paste(tmpdir, "/", dsname, "/", sep="") # get_dataset_dir(tar_gz_file)
#     metadata <- format_metadata(dataset_dir)
#     pheno_data <- format_pheno_data(metadata)
#     experiment_data <- format_experiment_data(metadata)
# # the next command writes out ExpressionSets
#     genefamilies_relab(pheno_data, experiment_data, dataset_dir)
#     marker_abundance(pheno_data, experiment_data, dataset_dir)
#     marker_presence(pheno_data, experiment_data, dataset_dir)
#     metaphlan_bugs_list(pheno_data, experiment_data, dataset_dir)
#     pathabundance_relab(pheno_data, experiment_data, dataset_dir)
#     pathcoverage(pheno_data, experiment_data, dataset_dir)
#     combine_metadata(tar_gz_file, metadata)
# }

make_data("data/test")
```

```{r}
# tar_gz_file <- "test"
# tmpdir <- cmg_tmpdir()
dataset_dir <- "~/curatedMetagenomicData/data/test"

    # dir.create(uploads_dir(path  = tmpdir))
    # dsname = gsub(".tar.gz", "", basename(tar_gz_file))
    # untar(tar_gz_file, exdir = tmpdir)
    # dataset_dir <- paste(tmpdir, "/", dsname, "/", sep="") # get_dataset_dir(tar_gz_file)
    metadata <- format_metadata(dataset_dir)
    pheno_data <- format_pheno_data(metadata)
    experiment_data <- format_experiment_data(metadata)
## the next command writes out ExpressionSets
    genefamilies_relab(pheno_data, experiment_data, dataset_dir)
    marker_abundance(pheno_data, experiment_data, dataset_dir)
    marker_presence(pheno_data, experiment_data, dataset_dir)
    metaphlan_bugs_list(pheno_data, experiment_data, dataset_dir)
    pathabundance_relab(pheno_data, experiment_data, dataset_dir)
    pathcoverage(pheno_data, experiment_data, dataset_dir)
    combine_metadata(tar_gz_file, metadata)
```







```{r warning=FALSE}
output_ind <- 63
output_path <- outputs$path[output_ind]
res <- read.table(gsutil_pipe(output_path), sep = "\t")
head(res)
```

```{r warning=FALSE}
output_name <- "humann_pathabundance_relab_counts.tsv"
output_ind <- which(outputs$file == output_name)
output_path <- outputs$path[output_ind]

relab_counts <- read.table(gsutil_pipe(output_path))
relab_counts
```

```{r warning=FALSE}
output_name <- "HSM7J4NY_pathabundance_relab.tsv"
output_ind <- which(outputs$file == output_name)
output_path <- outputs$path[output_ind]

HSM7J4NY_abundance <- read.table(gsutil_pipe(output_path), sep = "\t")
head(HSM7J4NY_abundance)
```
